### ISSUE TITLE
Bug: id token is erased when the server does not send a new idtoken on refresh

### ISSUE THREAD

**Issue by agardiol** [damienbod/angular-auth-oidc-client#1408]
**What Version of the library are you using?**
13.1.0

**Question**
I am trying to get refresh process working with a server that give no id token with the refresh answer:
```
{access_token: "ZEPGAiCD01MSG5qleO9gQzjD46c9D9", expires_in: 60, token_type: "Bearer",…}
access_token: "ZEPGAiCD01MSG5qleO9gQzjD46c9D9"
expires_in: 60
refresh_token: "hc3Crg96ehnOblKUImHAeGRgamPvf5"
scope: "openid read write"
token_type: "Bearer"
```

As explained in the documentation, I set the disableRefreshIdTokenAuthTimeValidation option to false and I manually call 
```
this._oidcSecurityService.forceRefreshSession().subscribe((result:any) => {
                  if (!result || result == null) {
                    console.error('Access token refresh process failed -> Logoff requested !');
                    ...
                  }
                });
```
I always get an error even if I can see in the log:
![image](https://user-images.githubusercontent.com/60869624/158601530-46959192-e6cc-4d00-91e4-76daf4e7a65e.png)

Note that access token expiration is set to 60s (for test) and id token expiration to 10s in the server.

**Comment by agardiol** [comment]
Thanks by advance !

**Comment by agardiol** [comment]
@damienbod 
Sorry to insist. Do you have any answer to this question ?
Thanks by advance.

**Comment by damienbod** [comment]
@agardiol This should work, I will test this again , doing a new release now

Greetings Damien

**Comment by agardiol** [comment]
Update for v14.0:
The behavior is still the same with the new version 14. The new access token is well updated in the authResult area in the session storage, but the returned object of `forceRefreshSession().subscribe((loginResponse: LoginResponse)` is still null even if the process succeed.
![image](https://user-images.githubusercontent.com/60869624/164005529-d53d44ed-cd5f-496a-8ed9-42aeb2ab7046.png)

**Comment by lwalejko** [comment]
I have also encountered this issue, here is the culprit: [https://github.com/damienbod/angular-auth-oidc-client/blob/9fc89c8a6ce8af4a2abd528c484326a6f0fa5f64/projects/angular-auth-oidc-client/src/lib/flows/callback-handling/history-jwt-keys-callback-handler.service.ts#L34](https://github.com/damienbod/angular-auth-oidc-client/blob/9fc89c8a6ce8af4a2abd528c484326a6f0fa5f64/projects/angular-auth-oidc-client/src/lib/flows/callback-handling/history-jwt-keys-callback-handler.service.ts#L34)

When `id_token` is missing in refresh response, existing `id_token` will be discarded.
@damienbod can this be patched by preserving current `id_token`, or will it break something else?

As a workaround without modifying package code I have patched `id_token` on save if it suddenly disappears, something like this:
```ts
import { Injectable } from '@angular/core';
import { AbstractSecurityStorage } from 'angular-auth-oidc-client';
import * as _ from 'lodash';

@Injectable({
  providedIn: 'root',
})
export class CustomSecurityStorage implements AbstractSecurityStorage {
  read(key: string) {
    return sessionStorage.getItem(key);
  }

  write(key: string, value: any): void {
    const currentDataRaw = sessionStorage.getItem(key);
    if (!currentDataRaw) {
      sessionStorage.setItem(key, value);
      return;
    }

    // Patch id_token when missing in token refresh response
    const currentDataParsed = JSON.parse(currentDataRaw) || {};
    const newDataParsed = JSON.parse(value) || {};

    const currentIdToken = _.get(currentDataParsed, 'authnResult.id_token');
    const newIdToken = _.get(newDataParsed, 'authnResult.id_token');
    const newAccessToken = _.get(newDataParsed, 'authnResult.access_token');

    // If id_token is currently present, but missing from refresh response, patch it with old value
    const newDataPatched =
      newAccessToken && currentIdToken && !newIdToken
        ? _.set(newDataParsed, 'authnResult.id_token', currentIdToken)
        : newDataParsed;

    sessionStorage.setItem(key, JSON.stringify(newDataPatched));
  }

  remove(key: string): void {
    sessionStorage.removeItem(key);
  }

  clear(): void {
    sessionStorage.clear();
  }
}
```

and then provide in app module
```ts
providers: [
    {
      provide: AbstractSecurityStorage,
      useClass: CustomSecurityStorage,
    },
  ],
```

Do you need help with the patch?

**Comment by damienbod** [comment]
Thanks for this, looks good and needed. Would you make a PR? @FabianGosebrink?

**Comment by bgerhards-vermeer** [comment]
Looks like this was given the investigate label back in May. A workaround has been given but it looks like there is some concern. Anything new on this issue or the others that are related to it @damienbod , @FabianGosebrink? I appreciate your time and I hope all is well!

**Comment by bgerhards** [comment]
Looks like this was given the investigate label back in May. A workaround has been given but it looks like there is some concern. Anything new on this issue or the others that are related to it @damienbod , @FabianGosebrink? I appreciate your time and I hope all is well!

**Comment by FabianGosebrink** [comment]
Is it normal or a misconfiguration that when refreshing the token, no ID token is being sent with?

**Comment by agardiol** [comment]
It depends on the type of the backend server. By example a standard Django configuration does not send ID token when refreshing session.

**Comment by bgerhards** [comment]
Another specific example is NetIQ: https://www.netiq.com/documentation/access-manager-45-developer-documentation/oauth-application-developer-guide/data/managing-tokens.html

**Comment by damienbod** [comment]
I will try to setup an IDP to test this. All the IDPs I use, return the id_token

**Comment by damienbod** [comment]
These 2 IDPs are not certified, iare these servers OIDC conform?

https://openid.net/certification/

**Comment by bgerhards** [comment]
Micro Focus Access Manager is apparently another name for it? Micro Focus owns NetIQ. If we are incorrect on Access Manager being another alias, please send correction.

**Comment by damienbod** [comment]
@bgerhards I will try to setup a OIDC Code flow with PKCE and refresh tokens using "Micro Focus Access Manager"

No promises when, never used this IDP before.

**Comment by FabianGosebrink** [comment]
Would the title "id token is erased when the server does not send a new idtoken on refresh" be a better fit?

**Comment by petersandor** [comment]
FYI This issue is a duplicate of https://github.com/damienbod/angular-auth-oidc-client/issues/1039#issuecomment-979481866 which was closed for no reason at all.

**Comment by jschutte27** [comment]
> @bgerhards I will try to setup a OIDC Code flow with PKCE and refresh tokens using "Micro Focus Access Manager"
> 
> No promises when, never used this IDP before.

@damienbod I can give you a clientID and a test user in our dev Microfocus access manager system if you would like to test it that way.

**Comment by damienbod** [comment]
@jschutte27 Yes please and the config to connect as well?

**Comment by jschutte27** [comment]
I sent you an email with the details.  Let me know if you don't get that or have questions.  I forgot to add the scopes to the email, but you can use "openid profile email roles".

**Comment by yasargil** [comment]
We have the same problem, token endpoint does not return an id token for a refresh token. After an automatic or manual refresh the session is broken.

**Comment by yasargil** [comment]
Seems like the id_token should be validated only once after the initial login. See: https://stackoverflow.com/a/25810348

**Comment by yasargil** [comment]
A simple fix would be to store the id_token like the access_token in a separate field.

**Comment by yasargil** [comment]
Following patch helps us when enableIdTokenExpiredValidationInRenew is false
```
diff --git a/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts b/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts
index 9aa0b6f7..4900c506 100644
--- a/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts
+++ b/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts
@@ -54,6 +54,11 @@ export class AuthStateService {
   ): void {
     this.loggerService.logDebug(currentConfig, `storing the accessToken '${accessToken}'`);
 
+    if (authResult.id_token != null) {
+      this.loggerService.logDebug(currentConfig, `storing the id token '${authResult.id_token}'`);
+      this.storagePersistenceService.write('authzIdTokenData', accessToken, currentConfig);
+    }
+
     this.storagePersistenceService.write('authzData', accessToken, currentConfig);
     this.persistAccessTokenExpirationTime(authResult, currentConfig);
     this.setAuthenticatedAndFireEvent(allConfigs);
diff --git a/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts b/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts
index 5865d5c2..ed3f19d7 100644
--- a/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts
+++ b/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts
@@ -5,6 +5,7 @@ import { BrowserStorageService } from './browser-storage.service';
 export type StorageKeys =
   | 'authnResult'
   | 'authzData'
+  | 'authzIdTokenData'
   | 'access_token_expires_at'
   | 'authWellKnownEndPoints'a
   | 'userData'
@@ -67,6 +68,7 @@ export class StoragePersistenceService {
 
   resetAuthStateInStorage(config: OpenIdConfiguration): void {
     this.remove('authzData', config);
+    this.remove('authzIdTokenData', config);
     this.remove('reusable_refresh_token', config);
     this.remove('authnResult', config);
   }
@@ -76,7 +78,7 @@ export class StoragePersistenceService {
   }
 
   getIdToken(config: OpenIdConfiguration): string {
-    return this.read('authnResult', config)?.id_token;
+    return this.read('authzIdTokenData', config);
   }
 
   getRefreshToken(config: OpenIdConfiguration): string {

```

**Comment by michal-husak** [comment]
> Following patch helps us when enableIdTokenExpiredValidationInRenew is false
> 
> ```
> diff --git a/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts b/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts
> index 9aa0b6f7..4900c506 100644
> --- a/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts
> +++ b/projects/angular-auth-oidc-client/src/lib/auth-state/auth-state.service.ts
> @@ -54,6 +54,11 @@ export class AuthStateService {
>    ): void {
>      this.loggerService.logDebug(currentConfig, `storing the accessToken '${accessToken}'`);
>  
> +    if (authResult.id_token != null) {
> +      this.loggerService.logDebug(currentConfig, `storing the id token '${authResult.id_token}'`);
> +      this.storagePersistenceService.write('authzIdTokenData', accessToken, currentConfig);
> +    }
> +
>      this.storagePersistenceService.write('authzData', accessToken, currentConfig);
>      this.persistAccessTokenExpirationTime(authResult, currentConfig);
>      this.setAuthenticatedAndFireEvent(allConfigs);
> diff --git a/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts b/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts
> index 5865d5c2..ed3f19d7 100644
> --- a/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts
> +++ b/projects/angular-auth-oidc-client/src/lib/storage/storage-persistence.service.ts
> @@ -5,6 +5,7 @@ import { BrowserStorageService } from './browser-storage.service';
>  export type StorageKeys =
>    | 'authnResult'
>    | 'authzData'
> +  | 'authzIdTokenData'
>    | 'access_token_expires_at'
>    | 'authWellKnownEndPoints'a
>    | 'userData'
> @@ -67,6 +68,7 @@ export class StoragePersistenceService {
>  
>    resetAuthStateInStorage(config: OpenIdConfiguration): void {
>      this.remove('authzData', config);
> +    this.remove('authzIdTokenData', config);
>      this.remove('reusable_refresh_token', config);
>      this.remove('authnResult', config);
>    }
> @@ -76,7 +78,7 @@ export class StoragePersistenceService {
>    }
>  
>    getIdToken(config: OpenIdConfiguration): string {
> -    return this.read('authnResult', config)?.id_token;
> +    return this.read('authzIdTokenData', config);
>    }
>  
>    getRefreshToken(config: OpenIdConfiguration): string {
> ```

This flag does not seem to work for me. The library just stops adding the access token to the header after the refresh call without id_token.

**Comment by yasargil** [comment]
This only works with the patch, without the token is lost after the refresh call.

**Comment by michal-husak** [comment]
I solved it with this interceptor :D

```typescript
@Injectable()
export class MissingIdTokenDuringRefreshCompensationInterceptor implements HttpInterceptor {

  intercept(request: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
    if (request.method === 'POST' && (request.body as string).includes('grant_type=refresh_token')) {
      return next.handle(request).pipe(
        tap((response: any) => {
          if (response.body) {
            response.body.id_token = response.body.id_token ?? response.body.access_token;
          }
        })
      );
    }
    return next.handle(request);
  }
}
```

as the id_token is not needed in my case anyway I can live with the access_token there as a placeholder.

**Comment by FabianGosebrink** [comment]
Basically we can fix this asking for an idtoken in the callback response, or am I missing something?

Like in `projects\angular-auth-oidc-client\src\lib\flows\callback-handling\history-jwt-keys-callback-handler.service.ts`

```
 callbackHistoryAndResetJwtKeys(
    callbackContext: CallbackContext,
    config: OpenIdConfiguration,
    allConfigs: OpenIdConfiguration[]
  ): Observable<CallbackContext> {
    if (!this.responseHasIdToken(callbackContext)) {
      const existingIdToken = this.storagePersistenceService.getIdToken(config);

      callbackContext.authResult.id_token = existingIdToken;
    }

    this.storagePersistenceService.write('authnResult', callbackContext.authResult, config);

    /* ... */
  }
```

Any thoughts?

**Comment by yasargil** [comment]
In our case the idtoken will be expired by then. So if there will be a check after that callback it will fail.

**Comment by michal-husak** [comment]
> Basically we can fix this asking for an idtoken in the callback response, or am I missing something?
> 
> Like in `projects\angular-auth-oidc-client\src\lib\flows\callback-handling\history-jwt-keys-callback-handler.service.ts`
> 
> ```
>  callbackHistoryAndResetJwtKeys(
>     callbackContext: CallbackContext,
>     config: OpenIdConfiguration,
>     allConfigs: OpenIdConfiguration[]
>   ): Observable<CallbackContext> {
>     if (!this.responseHasIdToken(callbackContext)) {
>       const existingIdToken = this.storagePersistenceService.getIdToken(config);
> 
>       callbackContext.authResult.id_token = existingIdToken;
>     }
> 
>     this.storagePersistenceService.write('authnResult', callbackContext.authResult, config);
> 
>     /* ... */
>   }
> ```
> 
> Any thoughts?

I was thinking that we could preserve the first original id_token in the sessionStorage after the token refresh call returns an answer without the id_token && if the `enableIdTokenExpiredValidationInRenew` is set to false.

**Comment by FabianGosebrink** [comment]
When the stored id token is expired and the refresh did not return a new idtoken - which idtoken should be stored then?

**Comment by yasargil** [comment]
The idtoken from the initial login event should be kept around as long as there will be no new token delivered and should be only validated for the initial login event and not for a refresh.

**Comment by michal-husak** [comment]
when it is expired and the flag is `enableIdTokenExpiredValidationInRenew = false` you don't really care anymore right?
I would just keep the old one there.

Or is there some other logic I am missing? 
In my workaround interceptor I just tricked it with the access_token which came from the refresh call.

Does the old id_token need to be not-expired in order to get `true` from the `isAuthenticated()` ?

**Comment by FabianGosebrink** [comment]
We are currently working on this.

**Comment by damienbod** [comment]
@jschutte27  Can you send me the server details again? We added a fix for this and would like to test with an IDP which does not return an id_token in the refresh

Thanks Damien

**Comment by jschutte27** [comment]
I sent it to your email address.  Let me know if you need anything else.

**Comment by damienbod** [comment]
@jschutte27 got it, thanks!

**Comment by damienbod** [comment]
We have a working version now of this using this configuration for the provided server. Just need some tests and docs now and we can finished the PR and release

```typescript
import { NgModule } from '@angular/core';
import { AuthModule, LogLevel } from 'angular-auth-oidc-client';

@NgModule({
  imports: [
    AuthModule.forRoot({
      config: {
        authority: '--idp--',
        redirectUrl: window.location.origin,
        postLogoutRedirectUri: window.location.origin,
        clientId: '--client_id--',
        scope: 'openid profile offline_access',
        responseType: 'code',
        silentRenew: true,
        useRefreshToken: true,
        ignoreNonceAfterRefresh: true, // this is required if the id_token is not returned
        allowUnsafeReuseRefreshToken: true, // this is required if the refresh token is not rotated
        enableIdTokenExpiredValidationInRenew: false, // required to refresh the browser if id_token is not updated after the first authentication
        autoUserInfo: false, // if the user endpoint is not supported
        logLevel: LogLevel.Debug,
      },
    }),
  ],
  exports: [AuthModule],
})
export class AuthConfigModule {}
```

**Comment by agardiol** [comment]
Hello @damienbod  and @FabianGosebrink 
Thanks for this fix.
I just updated my app with it (sorry I had no time to test it before) and unfortunately I got a new error. 
When the refresh process occurs, I get know this WARN messages:
![image](https://user-images.githubusercontent.com/60869624/203768455-7aca0e74-b4eb-4b4b-b148-72bb59c36ce5.png)
and the auth data in the session storage disappear (like in V13).
Here is my config: 
```
config: OpenIdConfiguration = {
      authority: environment.backend_url + "/o",
      redirectUrl: window.location.origin + (environment.production ? "/" + this.localeId : '') + "/backendReturn.html",  // Important
      clientId: environment.clientId,
      scope: 'openid read write offline_access',
      responseType: 'code',
      silentRenew: true,
      useRefreshToken: true,
      ignoreNonceAfterRefresh: true, // this is required if the id_token is not returned
      triggerRefreshWhenIdTokenExpired: false,    // required to refresh the browser if id_token is not updated after the first authentication
      renewTimeBeforeTokenExpiresInSeconds: 30,
      autoUserInfo: false, // if the user endpoint is not supported
      secureRoutes: [environment.backend_url + '/api/v0'],
      postLogoutRedirectUri: window.location.origin,
      //autoCleanStateAfterAuthentication: true,
      postLoginRoute: '/app',                               // Goto web app after successfull logon
      unauthorizedRoute: '/unauthorized',
      logLevel: (environment.production ? LogLevel.Warn : LogLevel.Debug)
      // Notes:
      // triggerAuthorizationResultEvent allows to emit an event instead of redirecting to redirectUrl (allow to save the current page and return to it after successfull login)
    };
```
Am I missing something ?
Can you also explain why the offline_access scope is now required ?
Thanks by advance !
